### PR TITLE
Improve `includes` in `core/variant`

### PR DESCRIPTION
**Pull request by Nazarwadim**
Similar to this #89568

**Benchmark**
SCU build time:
Before:
04:29.433
After:
04:26.880

Edit:
I made benchmarks for variants `.cpp` compille time.
Command (for linuxbsd):
```
time g++ -c -std=gnu++17 -fno-exceptions -Wctor-dtor-privacy -Wnon-virtual-dtor -Wplacement-new=1 -pipe -O3 -Wall -Wextra -Wwrite-strings -Wno-unused-parameter -Wshadow -Wno-misleading-indentation -Wno-type-limits -Walloc-zero -Wduplicated-branches -Wduplicated-cond -Wstringop-overflow=4 -Wattribute-alias=2 -Werror -DNDEBUG -DNO_EDITOR_SPLASH -DDISABLE_DEPRECATED -DSOWRAP_ENABLED -DTOUCH_ENABLED -DFONTCONFIG_ENABLED -DALSA_ENABLED -DALSAMIDI_ENABLED -DPULSEAUDIO_ENABLED -D_REENTRANT -DDBUS_ENABLED -DSPEECHD_ENABLED -DXKB_ENABLED -DJOYDEV_ENABLED -DUDEV_ENABLED -DLINUXBSD_ENABLED -DUNIX_ENABLED -D_FILE_OFFSET_BITS=64 -DX11_ENABLED -DLIBDECOR_ENABLED -DWAYLAND_ENABLED -DVULKAN_ENABLED -DRD_ENABLED -DGLES3_ENABLED -DCRASH_HANDLER_ENABLED -D_3D_DISABLED -DADVANCED_GUI_DISABLED -DBROTLI_ENABLED -DTHREADS_ENABLED -DCLIPPER2_ENABLED -DZSTD_STATIC_LINKING_ONLY -Ithirdparty/zstd -Ithirdparty/zlib -Ithirdparty/clipper2/include -Ithirdparty/brotli/include -Iplatform/linuxbsd -Ithirdparty/linuxbsd_headers/wayland -Ithirdparty/linuxbsd_headers -Iplatform/linuxbsd -I. <FILENAME.cpp>
```
| File                                   | Before  | After  |
|----------------------------------------|---------------------|--------------------|
| core/variant/array.cpp                | real 0m2.020s      | real 0m2.008s     |
|                                        | user 0m1.964s      | user 0m1.944s     |
|                                        | sys 0m0.080s       | sys 0m0.088s      |
| core/variant/callable.cpp             | real 0m0.829s      | real 0m0.825s     |
|                                        | user 0m0.778s      | user 0m0.759s     |
|                                        | sys 0m0.056s       | sys 0m0.070s      |
| core/variant/dictionary.cpp           | real 0m0.727s      | real 0m0.714s     |
|                                        | user 0m0.687s      | user 0m0.679s     |
|                                        | sys 0m0.037s       | sys 0m0.040s      |
| core/variant/variant.cpp              | real 0m2.753s      | real 0m2.694s     |
|                                        | user 0m2.698s      | user 0m2.633s     |
|                                        | sys 0m0.086s       | sys 0m0.092s      |
| core/variant/variant_call.cpp         | real 0m23.388s     | real 0m23.292s    |
|                                        | user 0m23.309s     | user 0m23.252s    |
|                                        | sys 0m0.329s       | sys 0m0.290s      |
| core/variant/variant_construct.cpp    | real 0m4.642s      | real 0m4.623s     |
|                                        | user 0m4.593s      | user 0m4.555s     |
|                                        | sys 0m0.114s       | sys 0m0.111s      |
| core/variant/variant_destruct.cpp     | real 0m0.491s      | real 0m0.403s     |
|                                        | user 0m0.470s      | user 0m0.367s     |
|                                        | sys 0m0.021s       | sys 0m0.039s      |
| core/variant/variant_op.cpp           | real 0m8.431s      | real 0m8.426s     |
|                                        | user 0m8.365s      | user 0m8.394s     |
|                                        | sys 0m0.165s       | sys 0m0.128s      |
| core/variant/variant_parser.cpp       | real 0m1.978s      | real 0m1.958s     |
|                                        | user 0m1.899s      | user 0m1.891s     |
|                                        | sys 0m0.095s       | sys 0m0.081s      |
| core/variant/variant_setget.cpp       | real 0m2.195s      | real 0m2.185s     |
|                                        | user 0m2.154s      | user 0m2.120s     |
|                                        | sys 0m0.072s       | sys 0m0.098s      |
| core/variant/variant_utility.cpp      | real 0m4.353s      | real 0m4.272s     |
|                                        | user 0m4.306s      | user 0m4.245s     |
|                                        | sys 0m0.103s       | sys 0m0.085s      |

### CODE REVIEW

**Comment by AThousandShips** [comment]
The benchmark you're giving isn't significant enough or relevant to this to be an argument either way I'd say, it's just misleading at best

**Comment by Nazarwadim** [comment]
> Without cache and cleaning each time?)

Without cache and with `--clean`

File: core/variant/variant.h [hunk 1]
```diff
@@ -31,7 +31,6 @@
 #ifndef VARIANT_H
 #define VARIANT_H
 
-#include "core/input/input_enums.h"
```
Comment (AThousandShips) [inline core/variant/variant.h:34]: This isn't valid, you can't forward declare these enums they're used

**Comment by AThousandShips** [comment]
An scu build won't really give any good indicators of this as they include a lot of headers and merge source files

But these are very core files and the includes here are important and can break things if not present

File: core/variant/variant.h (hunk 1 above)
Comment (Nazarwadim) [reply core/variant/variant.h:34]: Why, they are not used anywhere, and if they are, they are already included in the file that uses them

Comment (AThousandShips) [reply core/variant/variant.h:34]: Read the source of the header:
https://github.com/godotengine/godot/blob/fe01776f05b1787b28b4a270d53037a3c25f4ca2/core/variant/variant.h#L494-L507

**Comment by Nazarwadim** [comment]
> An scu build won't really give any good indicators of this as they include a lot of headers and merge source files

It is necessary to check both with SCU and without it, because most developers compile with SCU and it may be that there are no changes with SCU, or it may become worse.

**Comment by AThousandShips** [comment]
Yes, but you didn't test non-scu and the changes in the scu is (as we can see) negligible

**Comment by AThousandShips** [comment]
The individual build times for those files are so short that any measurement is unreliable, what matters is building the engine and exports

**Comment by Nazarwadim** [comment]
In general, my inclusions are correct because I used this tool [include-what-you-use](https://github.com/include-what-you-use/include-what-you-use) in current and #89568 PRs. In short, it is a wrapper over the clang compiler that outputs all these includes fixes to the console.
Like this:
```
core/variant/variant.h should add these lines:
#include "core/error/error_macros.h"         // for ERR_FAIL_COND_V_MSG
#include "core/math/math_defs.h"             // for real_t, Orientation, Side
#include "core/os/memory.h"                  // for memdelete, memnew
#include "core/string/string_name.h"         // for StringName
#include "core/templates/list.h"             // for List
#include "core/templates/safe_refcount.h"    // for SafeRefCount
#include "core/templates/vector.h"           // for Vector
#include "core/typedefs.h"                   // for _FORCE_INLINE_, _ALWAYS_...
#include <stdint.h>                          // for int64_t, uint32_t, uint64_t
enum class JoyAxis;
enum class JoyButton;
enum class Key;
enum class KeyLocation;
enum class MIDIMessage;
enum class MouseButton;

core/variant/variant.h should remove these lines:
- #include "core/input/input_enums.h"  // lines 34-34
- #include "core/os/keyboard.h"  // lines 54-54

The full include-list for core/variant/variant.h:
#include "core/error/error_macros.h"         // for ERR_FAIL_COND_V_MSG
#include "core/io/ip_address.h"              // for IPAddress
#include "core/math/aabb.h"                  // for AABB
#include "core/math/basis.h"                 // for Basis
#include "core/math/color.h"                 // for Color
#include "core/math/face3.h"                 // for Face3
#include "core/math/math_defs.h"             // for real_t, Orientation, Side
#include "core/math/plane.h"                 // for Plane
#include "core/math/projection.h"            // for Projection
#include "core/math/quaternion.h"            // for Quaternion
#include "core/math/rect2.h"                 // for Rect2
#include "core/math/rect2i.h"                // for Rect2i
#include "core/math/transform_2d.h"          // for Transform2D
#include "core/math/transform_3d.h"          // for Transform3D
#include "core/math/vector2.h"               // for Vector2
#include "core/math/vector2i.h"              // for Vector2i
#include "core/math/vector3.h"               // for Vector3
#include "core/math/vector3i.h"              // for Vector3i
#include "core/math/vector4.h"               // for Vector4
#include "core/math/vector4i.h"              // for Vector4i
#include "core/object/object_id.h"           // for ObjectID
#include "core/os/memory.h"                  // for memdelete, memnew
#include "core/string/node_path.h"           // for NodePath
#include "core/string/string_name.h"         // for StringName
#include "core/string/ustring.h"             // for String
#include "core/templates/list.h"             // for List
#include "core/templates/paged_allocator.h"  // for PagedAllocator
#include "core/templates/rid.h"              // for RID
#include "core/templates/safe_refcount.h"    // for SafeRefCount
#include "core/templates/vector.h"           // for Vector
#include "core/typedefs.h"                   // for _FORCE_INLINE_, _ALWAYS_...
#include "core/variant/array.h"              // for Array
#include "core/variant/callable.h"           // for Callable, Signal
#include "core/variant/dictionary.h"         // for Dictionary
#include <stdint.h>                          // for int64_t, uint32_t, uint64_t
class Object;  // lines 63-63
enum class JoyAxis;
enum class JoyButton;
enum class Key;
enum class KeyLocation;
enum class MIDIMessage;
enum class MouseButton;
struct MethodInfo;  // lines 66-66
struct PropertyInfo;  // lines 65-65
```

It has good logic and understands where to use Forward-declaration and where not to.

I read on their github that they used it for some Google program and improved its compilation speed by 30%.

**Comment by AThousandShips** [comment]
We don't rely on that tool, very aware of it, but it's very hit and miss and we have a generally very carefully managed include hierarchy, so this is not something that's entirely safe, so this needs to be evaluated on those grounds, and reorganizing includes is something that's fragile, and needs to be done when there's real reason to do so

The data you've presented here doesn't imply any benefit from this, and has risks in creating more fragile includes elsewhere, these classes are core-core, extremely central to how the engine works, so not something that's easy to restructure or mess with

**Comment by AThousandShips** [comment]
Well you haven't shown that they *do* improve build times? And you can't guarantee that they don't make the code more fragile, it doesn't break anything in the configurations you've build, and currently, but it might make includes for the engine more unpredictable and break other people's PRs, and currently you haven't shown any evidence it helps with anything :)

I haven't run a full non-scu build of release or anything, but my testing has shown that this doesn't improve anything, there's literally no change in build time between this and `master`:
Master:
* 00:03:29.209
* 00:03:26.735
This:
* 00:03:28.314
* 00:03:28.094

So I fail to see any improvement here?

**Comment by Nazarwadim** [comment]
Okay, now I'll test the build without scu, and if the master differs from mine by less than 0.5% growth, I'll close it.

**Comment by AThousandShips** [comment]
0.5% is a massively small limit, I'd say at least 1% is a normal variance

**Comment by AThousandShips** [comment]
So I ran once for each, just to get a starting line (which still shows the variance and limits we can expect) and it shows no actual improvement:
Before:
* 00:09:08.511

After:
* 00:08:54.581

Even if this were the averages (which matter relatively little, the variance is more important) you'd still just see a difference in 2.5%, which is insignificant, I can run a few more rounds, but the average isn't likely to be significantly better

**Comment by Nazarwadim** [comment]
I also turned off the turbo boost.
Before:
10;37.949
10:31:876
After:
10:19.530
10:24.283

**Comment by AThousandShips** [comment]
So that's a variance between the smallest gap of just 1.2%

**Pull request closed without merging**